BudgetSquirrel.Business/Tracking/FinalizeBudgetPeriodCommand.cs

```diff
@@ -0,0 +1,68 @@
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using BudgetSquirrel.Business.Auth;
+using BudgetSquirrel.Business.BudgetPlanning;
+using BudgetSquirrel.Business.Infrastructure;
+
+namespace BudgetSquirrel.Business.Tracking
+{
+    public class FinalizeBudgetPeriodCommand
+    {
+        private readonly IUnitOfWork unitOfWork;
+        private readonly Guid budgetId;
+        private readonly User currentUser;
+
+        public FinalizeBudgetPeriodCommand(IUnitOfWork unitOfWork, Guid budgetId, User currentUser)
+        {
+            this.unitOfWork = unitOfWork;
+            this.budgetId = budgetId;
+            this.currentUser = currentUser;
+        }
+
+        public async Task Run()
+        {
+            IRepository<BudgetPeriod> budgetPeriodRepository = this.unitOfWork.GetRepository<BudgetPeriod>();
+
+            BudgetPeriod budgetPeriod = await budgetPeriodRepository.GetAll().SingleOrDefaultAsync(y => y.BudgetId == budgetId);
+
+            GetRootBudgetQuery rootBudgetQuery = new GetRootBudgetQuery(unitOfWork, currentUser.Id);
+
+            Budget rootBudget = await rootBudgetQuery.Run();
+
+            if (!rootBudget.IsOwnedBy(this.currentUser))
+            {
+                throw new InvalidOperationException("Unauthorized");
+            }        
+
+            if (!CanBudgetBeFinalized(rootBudget))
+            {
+                throw new InvalidOperationException("Budget can't not be finalized be review your budget again.");
+            }
+
+            budgetPeriod.SetFinalizedDate();
+            budgetPeriodRepository.Update(budgetPeriod);
+
+            await this.unitOfWork.SaveChangesAsync();
+        }
+
+        private bool CanBudgetBeFinalized(Budget budget)
```

Comment:
This depends on a domain property of the budget. Most of this logic is actually defining that domain property. That domain property would be whether or not a budget is allocated correctly. The name that pops into mind is `IsFullyAllocated`. It would be basically this recursive method but move it to a getter with that name on the Budget model so that any code module can use it. It's a core domain concept, it doesn't only have to do with finalizing a budget period.

```
public class Budget
{
    ...
    
    public bool IsFullyAllocated
    {
        get
        {
            if (this.SetAmount != this.SubBudgetTotalPlannedAmount && this.SubBudgets.Count() != 0)
                return false;

            foreach (Budget subBudget in this.SubBudgets)
                if (!subBudget.IsFullyAllocated)
                    return false;

            return true;
        }
    }
}
```